### ISSUE TITLE
New spreadsheet uploads

### ISSUE THREAD

**Issue by jessyclubsoda** [club-soda/club-soda-guide#580]
We have some new spreadsheets to upload to the Guide

I have added them to this folder https://drive.google.com/drive/folders/1_WydEWEusLirceD1E1sT9vRrg8J_Jkea?usp=sharing

# Technical Checklist

+ [x] Improve/update the instructions for adding a `csv` file.
+ [x] Ensure that **no duplicates** are being insterted
+ [x] Ensure that **`email` addresses** are being inserted into `users` table.
+ [x] Run the script on **`staging`**
+ [x] Run script to import into **`production`**

**Comment by jessyclubsoda** [comment]
We've added another spreadsheet to the Google Drive

**Comment by nelsonic** [comment]
# Step 1 

Before importing any data, ensure that you have made a backup of the production database.
Follow the instructions in: https://github.com/club-soda/club-soda-guide/blob/master/dev-guide.md#using-real-data to backup, download and import the production data.

Having the latest production data on your `localhost` is _essential_

**Comment by nelsonic** [comment]
7428

![image](https://user-images.githubusercontent.com/194400/57765501-80219a00-76fd-11e9-9a5d-cf7f8605b051.png)



7486
![image](https://user-images.githubusercontent.com/194400/57783825-b9bacb00-7726-11e9-9c99-fb90ca1ceea1.png)


7486 - 7428 = 58

58 corresponds to the number of rows in the `.csv` file (_minus the header row_)

**Comment by nelsonic** [comment]
`venue_users` before insert: **310** rows
![image](https://user-images.githubusercontent.com/194400/57850496-f5f93480-77d5-11e9-886f-9c7421f79be1.png)

`venue_users` after the insert: **367** rows
![image](https://user-images.githubusercontent.com/194400/57850768-af580a00-77d6-11e9-83ce-cdb64fe6b55f.png)

This is consistent with the data we just imported because there are 58 rows but one of the rows does _not_ have an email address so we only expect 57 new entries in the `venue_users` table:
![image](https://user-images.githubusercontent.com/194400/57850885-01009480-77d7-11e9-9081-da6f7a91b133.png)

**Comment by RobStallion** [comment]
@nelsonic Now that we are creating users via import, it would make sense to email these users telling them that they have had an account created and provide them with a link to allow them to set a password. We should be able to follow a lot of the logic used in the signup controller for this

**Comment by nelsonic** [comment]
Users viewable in Admin UI:
![club-soda-users-bermondsey](https://user-images.githubusercontent.com/194400/57852281-a8cb9180-77da-11e9-9193-baee5d4f5c69.png)

**Comment by nelsonic** [comment]
https://club-soda-guide-staging.herokuapp.com/admin/users
BEFORE:
![staging-users](https://user-images.githubusercontent.com/194400/57860532-eafdce80-77ec-11e9-832d-a3231d6bb56e.png)

AFTER:
![staging-users-bermondsey](https://user-images.githubusercontent.com/194400/57860877-87c06c00-77ed-11e9-9aa3-2c225d44fe51.png)

**Comment by nelsonic** [comment]
@jessyclubsoda we have imported the data for Bermondsey Pub Company e.g:
https://www.clubsodaguide.com/venues/st-james-of-bermondsey-se16-4qz
![image](https://user-images.githubusercontent.com/194400/57864446-ceb16000-77f3-11e9-8524-1830fee62279.png)
![image](https://user-images.githubusercontent.com/194400/57864457-d40eaa80-77f3-11e9-9cd8-c850e1e80871.png)


Please confirm. 

> We are working on importing Craft Union next.
The reason it's taking a bit longer is we are adding more documentation 
so that someone `else` will be able to do the import in the future ... 
see: https://github.com/club-soda/club-soda-guide/pull/583/files?short_path=1a7fafb#diff-1a7fafb7721eeab931d7f1b292da20b0

**Comment by nelsonic** [comment]
Before: (_latest pubs are bermondsey..._)
![image](https://user-images.githubusercontent.com/194400/57866398-34531b80-77f7-11e9-9098-769fb98c1c26.png)

After:
![image](https://user-images.githubusercontent.com/194400/57872178-070c6a80-7803-11e9-8e41-6edee2c8df59.png)

7784 - 7486 = **298**

The Craft Union Spreadsheet (and corresponding CSV file) has **298** rows (_discounting the header row_)
![image](https://user-images.githubusercontent.com/194400/57872225-260afc80-7803-11e9-82cc-0e8237d95912.png)

<hr />

## Log of issues noted while processing the Craft Union .csv:

+ All websites for Craft Union are invalid because they do not contain `http://` or `https://` 
![image](https://user-images.githubusercontent.com/194400/57869385-d6c1cd80-77fc-11e9-99b6-3881ccce76ae.png)

+ URLs containing uppercase letters:
![image](https://user-images.githubusercontent.com/194400/57869707-84cd7780-77fd-11e9-8cc1-dfb5d4175154.png)

+ Space in URL:
![image](https://user-images.githubusercontent.com/194400/57870360-cf032880-77fe-11e9-92ec-dab63569b1ec.png)
![image](https://user-images.githubusercontent.com/194400/57870604-5781c900-77ff-11e9-8738-6c0a105042c9.png)

+ Twitter URL fails: (no forwardslash before username)
![image](https://user-images.githubusercontent.com/194400/57869321-babe2c00-77fc-11e9-85e9-9b166f918aa7.png)
+ Too many `t's` in the url (`htttp` ...)
![image](https://user-images.githubusercontent.com/194400/57870709-8ef07580-77ff-11e9-8c5e-1574ea009bba.png)

+ Incorrect postcode: `HR9 1EA` should be `HR9 7EA` (_according to Google_)
![image](https://user-images.githubusercontent.com/194400/57871548-784b1e00-7801-11e9-8d85-03c05bacf2ba.png)
![image](https://user-images.githubusercontent.com/194400/57871570-826d1c80-7801-11e9-9761-b070789d6d2b.png)

# _Next_: import the craft_union data into `staging`

**Comment by RobStallion** [comment]
Looking at the brewdog csv file for imports I have a few of the same issues as @nelsonic did above.

I am also getting an issue where one of the venues that we are trying to import already exists in app...

![image](https://user-images.githubusercontent.com/15571853/57873911-52c11300-7807-11e9-8e1d-c8a20b6e8c92.png)

![image](https://user-images.githubusercontent.com/15571853/57874020-8308b180-7807-11e9-862e-205594396149.png)

![image](https://user-images.githubusercontent.com/15571853/57873986-77b58600-7807-11e9-9da2-37cee879c5ac.png)

Going to skip this entry when running the import.

**Comment by nelsonic** [comment]
Imported 298 Craft Union Venues on Staging e.g:
![image](https://user-images.githubusercontent.com/194400/57874078-a9c6e800-7807-11e9-9918-00706e05db48.png)

https://club-soda-guide-staging.herokuapp.com/venues/brasserie-chalon-wa10-1rx
![image](https://user-images.githubusercontent.com/194400/57874749-37570780-7809-11e9-8d66-ab511c81f7b2.png)

The craft union spreadsheet does not have venue description or phone_number so that's why it looks "plain". in this instance the venue doesn't have Twitter 🐦 or Insta 📸 so it's a bit antisocial too ... 🙄 

But the import works. ✅ 

# _Next_: Importing data to Production 📥

**Comment by RobStallion** [comment]
Brewdog import to staging successful:

<img width="700" alt="brewdog" src="https://user-images.githubusercontent.com/15571853/57875452-e8aa6d00-780a-11e9-9b9c-b92943e45821.png">

**Comment by nelsonic** [comment]
@RobStallion PR ready for review: https://github.com/club-soda/club-soda-guide/pull/583 📦

**Comment by RobStallion** [comment]
![image](https://user-images.githubusercontent.com/15571853/57876864-4f7d5580-780e-11e9-95f1-bcd1181ec836.png)

brewdog uploaded to prod

**Comment by jessyclubsoda** [comment]
Thanks for all this guys! Much appreciated and thanks for logging the issues too

**Comment by nelsonic** [comment]
@jessyclubsoda when the drinks are spelled _exactly_ the way they are on the website, 
they are linked as per: https://www.clubsodaguide.com/venues/st-james-of-bermondsey-se16-4qz
If the drink is spelled differently the script is not "fuzzy matching" them to avoid "false positives".

**Comment by jessyclubsoda** [comment]
Has the script for updating the scores been run? There seems to be lots of venues with drinks but no score

**Comment by RobStallion** [comment]
@jessyclubsoda That has not been run yet. I am looking into the script at the moment to make sure that the venues are updated as expected.

**Comment by RobStallion** [comment]
@nelsonic while I was looking at the script to update the cs_score for venues I noticed that the Brewdog venues that were recently uploaded did not have drinks associated with them.

Nanny State, Raspberry Blitz, Punk AF are the three drinks for all Brewdog venues. The first 2 drinks are already in our drinks table...

![image](https://user-images.githubusercontent.com/15571853/58009692-e8e68900-7ae6-11e9-811d-132937681207.png)

![image](https://user-images.githubusercontent.com/15571853/58009652-d8cea980-7ae6-11e9-887d-89801532c5fe.png)

![image](https://user-images.githubusercontent.com/15571853/58009758-09aede80-7ae7-11e9-95d7-3345fb823d4b.png)

Punk AF is not.

After running the script to import the venues the correct number of venues are imported into the database (as shown in the screenshot above), however the drinks are not being associated with the venues (punk AF is also not created but I don't think that this is meant to happen at this stage as typos in drinks would create new, unwanted drinks.)

<img src="https://user-images.githubusercontent.com/15571853/58010206-dd479200-7ae7-11e9-853e-803b5b95a2e5.png" width=300px />

After doing a bit of looking into why this was the case I realised that the brand name is normally being taken from the drink field in the csv...

Some of the drinks belonging to Cross Keys Hotel...
![image](https://user-images.githubusercontent.com/15571853/58012007-6613fd00-7aeb-11e9-8ca6-d3595008e6b7.png)

As you can see the drinks all start with the brand name. 

In the case of Brewdog however, their drinks do not contain the brand name...
![image](https://user-images.githubusercontent.com/15571853/58012247-f81c0580-7aeb-11e9-95c1-7abd50e27bd7.png)

This is why the drinks are not being associated and why when I was testing the script to update cs_score on Brewdog venues, none were being updated. 

Once the drinks are added to the venue, the cs_score script appears to work as expect and updates the venue's score.

**Comment by RobStallion** [comment]
This appears to only have effected Brewdog in the latest venue uploads. Going to look into ways of updating the drinks for the Brewdog venues. Once this has been done I will run the script to update the venue cs scores

**Comment by jessyclubsoda** [comment]
I updated the Brewdog venues myself because I realised I had mucked up the spreadsheets

![image](https://user-images.githubusercontent.com/49029914/58013365-6235aa00-7aee-11e9-92db-53c92033540a.png)

Punk AF does exist on the site too

![image](https://user-images.githubusercontent.com/49029914/58013382-6b267b80-7aee-11e9-95d5-b0d1d0800b45.png)

It seemed to mainly be the Craft Union pubs which had drinks associated with them but no score

**Comment by RobStallion** [comment]
@jessyclubsoda Thanks for that. I'll run the script to update the cs_scores in a few minutes then 👍 

Also thanks for pointing out that Punk AF exists. Not quite sure how I missed that earlier 😖

**Comment by RobStallion** [comment]
Made a [pr](https://github.com/club-soda/club-soda-guide/pull/586) to fix a small issue with the cs score script. 

Taking a back up of staging and prod. Once this pr has been merged I will run the script.

**Comment by nelsonic** [comment]
@RobStallion thanks for updating the script. ✅ (PR Merged)

**Comment by RobStallion** [comment]
Pr has been merged. Going to run the script now

**Comment by RobStallion** [comment]
Staging db copy - no. of venues pre script...
![image](https://user-images.githubusercontent.com/15571853/58021601-3ec92a00-7b03-11e9-9b61-9cbf11eb26fc.png)

post script
![image](https://user-images.githubusercontent.com/15571853/58021702-8c459700-7b03-11e9-972c-990020294310.png)

Looks like one venue was updated but the rest stopped due to an error...
```sh
** (Ecto.CastError) expected params to be a :map, got: `%CsGuide.Accounts.User{__meta__: #Ecto.Schema.Metadata<:loaded, "users">, admin: false, deleted: false, email: :error, email_hash: <<99, 10, 127, 70, 137, 248, 234, 207, 88, 100, 185, 113, 131, 79, 219, 205, 192, 27, 6, 109, 109, 139, 80, 110, 168, 141, 21, 250, 92, 26, 6, 160>>, email_plaintext: nil, entry_id: "5534889e-c48a-44c5-8ddc-8cbad147c903", id: 54, inserted_at: ~N[2019-05-16 14:13:56.000563], password: nil, password_confirmation: nil, password_reset_token: nil, password_reset_token_expiry: nil, role: :venue_admin, updated_at: ~N[2019-05-16 14:13:56.000624], venues: #Ecto.Association.NotLoaded<association :venues is not loaded>, verified: nil}`
    (ecto) lib/ecto/changeset.ex:409: Ecto.Changeset.cast/4
    (cs_guide) lib/cs_guide/accounts/user.ex:34: CsGuide.Accounts.User.changeset/2
    (ecto) lib/ecto/changeset.ex:736: anonymous fn/4 in Ecto.Changeset.on_cast_default/2
    (ecto) lib/ecto/changeset/relation.ex:122: Ecto.Changeset.Relation.do_cast/5
    (ecto) lib/ecto/changeset/relation.ex:312: Ecto.Changeset.Relation.map_changes/8
    (ecto) lib/ecto/changeset.ex:715: Ecto.Changeset.cast_relation/4
    (cs_guide) lib/cs_guide/resources/venue.ex:84: CsGuide.Resources.Venue.changeset/2
    (cs_guide) lib/cs_guide/resources/venue.ex:137: CsGuide.Resources.Venue.update/2
    priv/repo/update_cs_score.exs:25: CsGuide.UpdateCsScore.update_cs_score/2
    (elixir) lib/enum.ex:1940: Enum."-reduce/3-lists^foldl/2-0-"/3
    priv/repo/update_cs_score.exs:38: (file)
    (elixir) lib/code.ex:767: Code.require_file/2
    (mix) lib/mix/tasks/run.ex:147: Mix.Tasks.Run.run/5
    (mix) lib/mix/tasks/run.ex:86: Mix.Tasks.Run.run/1
    (mix) lib/mix/task.ex:331: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:79: Mix.CLI.run_task/2
```

**Comment by RobStallion** [comment]
This error is related to the users associated with a venue. As the users associated with the venue do not need to be updated in this script I am just going to remove them from the venue params before we run the update function. This should resolve the issue.

**Comment by RobStallion** [comment]
Have managed to run the script with no errors on the copies of staging and prod. Will open a new pr to push the latest changes and then should be able to run the script without error.

**Comment by RobStallion** [comment]
taking a new backup of the staging db now.

Once I have the copy I will run the command locally using the staging data. If that works, then I will be running the `mix run priv/repo/update_cs_score.exs` command in the heroku bash terminal once that is done. 

If all works as expected then I will repeat the process with the production db

**Comment by RobStallion** [comment]
Ran the command locally and all worked. Running on staging now

**Comment by RobStallion** [comment]
Before running the script 

![image](https://user-images.githubusercontent.com/15571853/58088373-7515b080-7bba-11e9-9f81-0f3e00d7f93e.png)

**Comment by RobStallion** [comment]
Got an error...

```
** (Ecto.CastError) expected params to be a :map, got: `%CsGuide.Accounts.User{__meta__: #Ecto.Schema.Metadata<:loaded, "users">, admin: false, deleted: false, email: "019157@BermondseyPubCo.com", email_hash: <<99, 10, 127, 70, 137, 248, 234, 207, 88, 100, 185, 113, 131, 79, 219, 205, 192, 27, 6, 109, 109, 139, 80, 110, 168, 141, 21, 250, 92, 26, 6, 160>>, email_plaintext: nil, entry_id: "5534889e-c48a-44c5-8ddc-8cbad147c903", id: 54, inserted_at: ~N[2019-05-16 14:13:56.000563], password: nil, password_confirmation: nil, password_reset_token: nil, password_reset_token_expiry: nil, role: :venue_admin, updated_at: ~N[2019-05-16 14:13:56.000624], venues: #Ecto.Association.NotLoaded<association :venues is not loaded>, verified: nil}`
    (ecto) lib/ecto/changeset.ex:409: Ecto.Changeset.cast/4
    (cs_guide) lib/cs_guide/accounts/user.ex:34: CsGuide.Accounts.User.changeset/2
    (ecto) lib/ecto/changeset.ex:736: anonymous fn/4 in Ecto.Changeset.on_cast_default/2
    (ecto) lib/ecto/changeset/relation.ex:122: Ecto.Changeset.Relation.do_cast/5
    (ecto) lib/ecto/changeset/relation.ex:312: Ecto.Changeset.Relation.map_changes/8
    (ecto) lib/ecto/changeset.ex:715: Ecto.Changeset.cast_relation/4
    (cs_guide) lib/cs_guide/resources/venue.ex:84: CsGuide.Resources.Venue.changeset/2
    (cs_guide) lib/cs_guide/resources/venue.ex:137: CsGuide.Resources.Venue.update/2
```

This is the same error I had locally yesterday which makes me think that the latest version of the script is not on staging. Looking into this now.

**Comment by RobStallion** [comment]
I can the `cat` command in the heroku bash terminal and got the following 

```sh
~ $ cat priv/repo/update_cs_score.exs
```
```ex
defmodule CsGuide.UpdateCsScore do
  alias CsGuide.Resources.{Venue, Drink}
  import Ecto.Query, only: [from: 2, subquery: 1]

  def update_cs_score(venue, acc) do
    query = fn s, m ->
      sub =
        from(mod in Map.get(m.__schema__(:association, s), :queryable),
          distinct: mod.entry_id,
          order_by: [desc: :updated_at],
          select: mod
        )

      from(m in subquery(sub), where: not m.deleted, select: m)
    end

    calculatedScore =
      CsGuide.Resources.CsScore.calculateScore(
        venue.drinks
        |> CsGuide.Repo.preload(drink_types: query.(:drink_types, Drink)),
        venue.num_cocktails
      )

    if venue.cs_score != calculatedScore && length(venue.drinks) != 0 do
      Venue.update(
        venue,
        venue |> Map.from_struct() |> Map.merge(%{cs_score: calculatedScore})
      )
      [{venue.venue_name, venue.entry_id} | acc]
    else
      acc
    end
  end
end
```

This is not the latest version of the script, as you can see if you navigate to [the file](https://github.com/club-soda/club-soda-guide/blob/master/priv/repo/update_cs_score.exs) on master.

**Comment by RobStallion** [comment]
Current version on staging is behind the current master.

Staging version
![image](https://user-images.githubusercontent.com/15571853/58089279-43054e00-7bbc-11e9-9553-daaae3205a5c.png)

Master versions
![image](https://user-images.githubusercontent.com/15571853/58089344-6f20cf00-7bbc-11e9-8a3c-6669241693fe.png)

Staging should be `5659...` but it looks like the automatically deploy failed for some reason. Looking into this now

**Comment by RobStallion** [comment]
The travis build failed yesterday. Appears to have timed out. I restarted the build and all seems to working as expected now...

travis
![image](https://user-images.githubusercontent.com/15571853/58089686-5533bc00-7bbd-11e9-94a5-88bb6e9dbe73.png)

Heroku is now building the latest version...
![image](https://user-images.githubusercontent.com/15571853/58089672-4d741780-7bbd-11e9-8f25-baa5dd184bbd.png)

**Comment by RobStallion** [comment]
Staging is now up to date ...

![image](https://user-images.githubusercontent.com/15571853/58089767-83190080-7bbd-11e9-8c0c-f6a262523949.png)

Re-running script

**Comment by RobStallion** [comment]
Staging update has been run successfully. Moving onto prod.

**Comment by RobStallion** [comment]
pre script...
![image](https://user-images.githubusercontent.com/15571853/58093203-f292ee00-7bc5-11e9-9dcb-d9bbdb9eb08b.png)

post script...
![image](https://user-images.githubusercontent.com/15571853/58093394-56b5b200-7bc6-11e9-998b-4adce46e8642.png)

update cs score script seems to have worked as expected on prod.
